### DISCUSSION TITLE
Mirror screen output for LCD-Projector use

### DISCUSSION THREAD

**Discussion by Demuos** [Bodmer/TFT_eSPI#2201] [Q&A - General]
Hello,
I have build a Projector based on the T-QT ESP32-S3 dev module with a 0.85" LCD https://github.com/Xinyuan-LilyGO/T-QT . The driver is a GC9107 [DataSheet](https://github.com/Xinyuan-LilyGO/T-QT/blob/main/doc/GC9107%20DataSheet%20V1.2.pdf) which is compatible with the implemented ST7789. The Display works great with the TFT_eSPI, but because of the light projection through the LCD the actual output of the Projector ist mirrored. Normal output of the LCD is correct. Is there a way to mirror the actual output? I´m happy for any help :)

EDIT:
carefully reading the DataSheet for the GC9107 there is a Frame Data Write Direction Definition were X-invert can be defined. The Question is now how I can use this Definition for my need and where I have to define this. Initialisation of the ST7789 Driver?
![image](https://user-images.githubusercontent.com/91437091/203802370-830f8258-cddc-4ea2-83ae-fc7094ff0b7f.png)
EDIT2:
I rethought the Problem and just installed the lcd in reverse which solved the problem.

**Comment by Bodmer** [comment]
Good solution!

There is a software approach which I will describe for completeness.

The MV, MX and MY bits are part of a control register (MADCTL) that also controls things like the RGB or GRB colour order for the pixels. Some displays also have an MH bit.

Typical allocations for the bits are in the display xxx_defines.h file. The bit allocations for the ST7789 are:
```
// Flags for TFT_MADCTL
#define TFT_MAD_MY  0x80
#define TFT_MAD_MX  0x40
#define TFT_MAD_MV  0x20
#define TFT_MAD_ML  0x10
#define TFT_MAD_RGB 0x00
#define TFT_MAD_BGR 0x08
#define TFT_MAD_MH  0x04
#define TFT_MAD_SS  0x02
#define TFT_MAD_GS  0x01
```
These bits are manipulated to implement rotations for setRotation(...) in the xxx_Rotation.h file for each display type. [Example here](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_Drivers/ST7789_Rotation.h).

Lines [like this one](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_Drivers/ST7789_Rotation.h#L68) set (or do not set) these mirror bits.

So it is  a case of deciding the rotation to be used and changing the rotation code. Or writing to the register in the sketch with the appropriate bits set or cleared. This will require some experimentation to get it right.
